lib/active_merchant/billing/gateways/quickbooks.rb

```diff
@@ -0,0 +1,312 @@
+        'PMT-5001' => STANDARD_ERROR_CODE[:card_declined],      # Merchant does not support given payment method
+
+        # System Error
+        'PMT-6000' => STANDARD_ERROR_CODE[:processing_error],   # A temporary Issue prevented this request from being processed.
+      }
+
+      FRAUD_WARNING_CODES = ['PMT-1000','PMT-1001','PMT-1002','PMT-1003']
+
+      def initialize(options = {})
+        requires!(options, :consumer_key, :consumer_secret, :access_token, :token_secret, :realm)
+        @consumer_key = options[:consumer_key]
+        @consumer_secret = options[:consumer_secret]
+        @access_token = options[:access_token]
+        @token_secret = options[:token_secret]
+        @realm = options[:realm]
+        super
+      end
+
+      def purchase(money, payment, options = {})
+        post = {}
+        add_amount(post, money, options)
+        add_charge_data(post, payment, options)
+        post[:capture] = "true"
+
+        commit(ENDPOINT, post)
+      end
+
+      def authorize(money, payment, options = {})
+        post = {}
+        add_amount(post, money, options)
+        add_charge_data(post, payment, options)
+        post[:capture] = "false"
+
+        commit(ENDPOINT, post)
+      end
+
+      def capture(money, authorization, options = {})
+        capture_uri = "#{ENDPOINT}/#{CGI.escape(authorization)}/capture"
+
+        commit(capture_uri)
+      end
+
+      def refund(money, authorization, options = {})
+        post = {}
+        post[:amount] = money.is_a?(String) ? money : amount(money)
+        refund_uri = "#{ENDPOINT}/#{CGI.escape(authorization)}/refunds"
+
+        commit(refund_uri, post)
+      end
+
+      def void(authorization, options = {})
+        MultiResponse.run do |r|
+          r.process { auth_object_from(authorization) }
+          if amount = r.params['amount']
+            r.process { refund(amount, authorization, options) }
+          end
+        end.responses.last
+      end
+
+      def verify(credit_card, options = {})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(1.00, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((realm=\")\w+), '\1[FILTERED]').
+          gsub(%r((oauth_consumer_key=\")\w+), '\1[FILTERED]').
+          gsub(%r((oauth_nonce=\")\w+), '\1[FILTERED]').
+          gsub(%r((oauth_signature=\")[a-zA-Z%0-9]+), '\1[FILTERED]').
+          gsub(%r((oauth_token=\")\w+), '\1[FILTERED]').
+          gsub(%r((\"card\":{\"number\":\")\d+), '\1[FILTERED]').
+          gsub(%r((\"cvc\":\")\d+), '\1[FILTERED]')
+      end
+
+      private
+
+      def add_charge_data(post, payment, options = {})
+        add_payment(post, payment, options)
+        add_address(post, options)
+        add_context(options[:context]) if options[:context]
```

Comment:
Not a fan of having a gateway-specific, nested `:context` option. Seems like it takes `:tax` and `:recurring`? I'd put them both at the top level of `options` if we want to support them, vs. nesting them.